// @refinedev/inferencer/dist/headless.js
var Ar=Object.create;var ue=Object.defineProperty;var Rr=Object.getOwnPropertyDescriptor;var wr=Object.getOwnPropertyNames;var Sr=Object.getPrototypeOf,Tr=Object.prototype.hasOwnProperty;var Nr=(e,r)=>{for(var s in r)ue(e,s,{get:r[s],enumerable:!0})},We=(e,r,s,o)=>{if(r&&typeof r=="object"||typeof r=="function")for(let p of wr(r))!Tr.call(e,p)&&p!==s&&ue(e,p,{get:()=>r[p],enumerable:!(o=Rr(r,p))||o.enumerable});return e};var R=(e,r,s)=>(s=e!=null?Ar(Sr(e)):{},We(r||!e||!e.__esModule?ue(s,"default",{value:e,enumerable:!0}):s,e)),Lr=e=>We(ue({},"__esModule",{value:!0}),e);var nt={};Nr(nt,{HeadlessCreateInferencer:()=>he,HeadlessCreateRenderer:()=>Me,HeadlessEditInferencer:()=>xe,HeadlessEditRenderer:()=>_e,HeadlessInferencer:()=>tt,HeadlessListInferencer:()=>ge,HeadlessListRenderer:()=>Ke,HeadlessShowInferencer:()=>fe,HeadlessShowRenderer:()=>ze});module.exports=Lr(nt);var pe=R(require("react")),Cr=require("@refinedev/core");var _=R(require("react")),He=require("@refinedev/core");var Vr=/(image|photo|avatar|url|logo|cover|thumbnail|icon)/i,$e=(e,r)=>{let s=["name","label","title","count","content","username","nickname","login","firstName","lastName","url"];Vr.test(e)&&s.unshift("url");let o=s.find(p=>p in r&&!!r[p]);if(o)return o==="firstName"&&"lastName"in r?["firstName","lastName"]:o};var Ze=require("@refinedev/core");function Ie(e,r,s){if(r)return r;let o=s==null?void 0:s.find(c=>c.name===e),p=(0,Ze.pickNotDeprecated)(o==null?void 0:o.meta,o==null?void 0:o.options);return p!=null&&p.dataProviderName?p.dataProviderName:"default"}var ae=e=>{var r;return(r=e==null?void 0:e.options)==null?void 0:r.dataProviderName};var be=e=>{let r=[...e].filter(Boolean).sort((s,o)=>(o.priority||0)-(s.priority||0))[0];return r||null};var Ge=R(require("prettier/standalone")),Ye=R(require("prettier/parser-typescript")),ke=e=>{try{return Ge.default.format(e,{parser:"typescript",plugins:[Ye.default],arrowParens:"always",trailingComma:"all",semi:!0,tabWidth:4,printWidth:80})}catch(r){return console.warn("Formatting error in inferencer",r),e}};var ve=e=>e.replace(/ {2,}/g," ").replace(/\n{2,}/g,`
`);var Ce=/(-id|-ids|_id|_ids|Id|Ids|ID|IDs)(\[\])?$/,Qe=(e,r)=>{let s=Ce.test(e),o=typeof r=="string"||typeof r=="number",p=Array.isArray(r)&&r.every(c=>typeof c=="string"||typeof c=="number");return s&&o||s&&p?{key:e,relation:!0,multiple:!!p,type:"relation",priority:1}:!1};var H=e=>e.replace(Ce,"");var f=e=>H(e).replace(/([a-z])([A-Z])/g,"$1 $2").replace(/_/g," ").replace(/-/g," ").split(" ").map(u=>u.charAt(0).toUpperCase()+u.slice(1)).join(" ");var Fe=e=>e.replace(/export\s+(const|let|var|type|interface|function|class)\s+(\w+)\s*(=|:)\s*/g,"$1 $2 = ").split(`
`).filter(p=>!p.trim().startsWith("export default")).join(`
`);var Xe=/import(?:(?:(?:[ \n\t]+([^ *\n\t\{\},]+)[ \n\t]*(?:,|[ \n\t]+))?([ \n\t]*\{(?:[ \n\t]*[^ \n\t"'\{\}]+[ \n\t]*,?)+\})?[ \n\t]*)|[ \n\t]*\*[ \n\t]*as[ \n\t]+([^ \n\t\{\}]+)[ \n\t]+)from[ \n\t]*(?:['"])([^'"\n]+)(?:['"])(?:;?)/g,Er=/import[ \n\t](?:['"])([^'"\n]+)(?:['"])(?:;?)/g,Pr=/((?:\w|\s|_)*)( as )((?:\w|\s|_)*)( |,)?/g,Ae=(e,r={})=>{let s=e.matchAll(Xe),o=new Set;for(let p of s){let[,c,u,h,m]=p;if(m in r){let x=r[m];if(m==="react")continue;c&&o.add(`const { default: ${c} } = ${x};`),u&&o.add(`const${u.replace(Pr,"$1: $3$4")} = ${x};`),h&&o.add(`const ${h} = ${x};`)}}return ve(`
    ${Array.from(o).join(`
`)}
    ${e.replace(Xe,"").replace(Er,"")}
    `)};var se=R(require("pluralize"));var Re=(e,r)=>e?r.find(o=>{let p=H(e.key);return o.name===p||o.name===(se.default.isPlural(p)?p:se.default.plural(p))||o.name===(se.default.isSingular(p)?p:se.default.singular(p))}):void 0;var Je=R(require("pluralize"));var M=(e,r,s="Inferred")=>{let o=e.replace(/[^a-zA-Z0-9]/g," "),p=Je.default.singular(o),u=f(p).replace(/ /g,"")||s,h=r.charAt(0).toUpperCase()+r.slice(1);return/^[a-zA-Z]/.test(u)?`${u}${h}`:`${h}${u}`};var Dr=e=>e?Array.isArray(e)?e.join(""):typeof e=="string"?e:typeof e=="object"||typeof e=="function"?e.toString():typeof e=="number"?String(e):typeof e=="boolean"&&e?"true":e:"",g=(e,...r)=>e.reduce((o,p,c)=>`${o}${p}${Dr(r[c])}`,"").trim();var we=(e,r)=>`
    ${e!=null?e:""}
    
    render(typeof ${r} !== "undefined" ? <${r} /> : <></>);
    `;var Se=e=>e==null?void 0:e.replace(/(\/\* hidden-start \*\/)(.|\n)*?(\/\* hidden-end \*\/)/g,"");var Or=/^[a-zA-Z_$][a-zA-Z_$0-9]*$/,Te=e=>Or.test(e);function Ne(e,r,s){let o=`${e}`;return r&&(o+="?.",Te(r)?o+=r:o+=`['${r}']`),s&&(o+="?.",Te(s)?o+=s:o+=`['${s}']`),o}function jr(e,r,s){return(s!=null?s:[]).map(p=>Ne(e,r,p))}var l=(e,r,s,o=' + " " + ')=>Array.isArray(s)?o?jr(e,r,s).join(o):Ne(e,r,s[0]):Ne(e,r,s),P=(e,r,s)=>{let o=e;return r&&(o+=`.${r}`),s&&(Array.isArray(s)?o+=`.${s[0]}`:o+=`.${s}`),o};var B=e=>{let r=e.reduce((o,[p,c,u])=>(o[c]||(o[c]=[]),o[c].includes(p)||(u?o[c].push([p,!0]):o[c].push(p)),o),{});return Object.entries(r).map(([o,p])=>{let c=p.find(x=>Array.isArray(x)&&x[1]),u=p.filter(x=>typeof x=="string"),h=c?`${c[0]}`:"",m=u.length?`{ ${u.join(", ")} }`:"";return`import ${h}${h&&m?", ":""}${m} from "${o}";`})};var Le=R(require("pluralize")),D=(e="")=>Le.default.isSingular(e)?e:Le.default.singular(e);var Ve=R(require("pluralize")),z=(e="")=>Ve.default.isPlural(e)?e:Ve.default.plural(e);var Hr=/^id$/i,ie=e=>Hr.test(e);var ce=e=>e.relationInfer&&e.relationInfer.accessor?Array.isArray(e.relationInfer.accessor)?`optionLabel: "${e.relationInfer.accessor[0]}",`:e.relationInfer.accessor!=="title"?`optionLabel: "${e.relationInfer.accessor}",`:"":"";var q=e=>{};var A=(e,r="")=>{if(Number.isInteger(Number(e)))return r.charAt(0).toLowerCase()+r.slice(1)+e;let o=e.replace(/^\d+/,"").replace(/[^a-zA-Z0-9]/g,"-"),p=f(o).replace(/ /g,"");return`${p.charAt(0).toLowerCase()+p.slice(1)}${r.charAt(0).toUpperCase()+r.slice(1)}`};var er=e=>{let r=(s,o,p,c=r)=>{let u=e.map(m=>m(s,o,p,c));return be(u)};return r};var rr=e=>(s,o,p,c,u)=>e.reduce((h,m)=>m(h,o,p,c,u),s);var tr=(e,r,s,o)=>{let p=Array.isArray(r),c=Array.isArray(r)&&r.every(u=>typeof u=="string"||typeof u=="number");if(p){if(!c){let h=o(e,r[0],s,o);return h?{...h,key:e,multiple:!0,priority:1}:!1}let u=o(e,r[0],s,o)||{type:"string"};return{key:e,multiple:!0,priority:1,type:u==null?void 0:u.type}}return!1};var nr=(e,r)=>typeof r=="boolean"?{key:e,type:"boolean"}:!1;var Ee=R(require("dayjs")),zr=/(_at|_on|At|On|AT|ON)(\[\])?$/,Kr=["/",":","-","."],or=(e,r)=>{let s=zr.test(e)&&(0,Ee.default)(r).isValid(),o=typeof r=="string"&&(0,Ee.default)(r).isValid();return typeof r=="string"&&Kr.some(c=>r.includes(c))&&(s||o)?{key:e,type:"date",priority:1}:!1};var Mr=/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,ar=(e,r)=>typeof r=="string"&&Mr.test(r)?{key:e,type:"email",priority:1}:!1;var _r=/\.(gif|jpe?g|tiff?|png|webp|bmp|svg)$/i,sr=(e,r)=>typeof r=="string"&&_r.test(r)?{key:e,type:"image",priority:2}:!1;var ir=(e,r)=>typeof r>"u"||r===null?null:!1;var cr=(e,r)=>{let s=typeof r=="string"&&r.length>0,o=!isNaN(r);return s&&o||typeof r=="number"?{key:e,type:"number"}:!1};var Br=/id$/i,pr=(e,r,s,o)=>{let p=r!==null,c=!Array.isArray(r);if(p&&c&&typeof r=="object"){if(Object.keys(r).length===1&&Br.test(Object.keys(r)[0]))return{key:e,type:"relation",relation:!0,accessor:"id",priority:1};let m=$e(e,r);if((m?m.length>0:!1)&&m){let F=m&&Array.isArray(m)?m[0]:m,k=o(F,r[F],r,o);if(k){let $=Array.isArray(m)?m.map(I=>k.accessor?`${I}.${Array.isArray(k.accessor)?k.accessor[0]:k.accessor}`:I):k.accessor?Array.isArray(k.accessor)?`${m}.${k.accessor[0]}`:`${m}.${k.accessor}`:m;return{...k,fieldable:!0,key:e,accessor:$,priority:1}}}return{key:e,fieldable:!1,type:"object",priority:1}}return!1};var ur=(e,r)=>typeof r=="string"&&r.length>100?{key:e,type:"richtext",priority:1}:!1;var lr=(e,r)=>typeof r=="string"?{key:e,type:"text"}:!1;var qr=/^(https?|ftp):\/\/(-\.)?([^\s/?\.#-]+\.?)+(\/[^\s]*)?$/i,mr=(e,r)=>typeof r=="string"&&qr.test(r)?{key:e,type:"url",priority:1}:!1;var dr=[tr,nr,or,ar,sr,ir,cr,pr,Qe,ur,lr,mr];var yr=(e,r,s,o)=>e.map(c=>{if(!c.relation&&(c.type==="text"||c.type==="richtext"||c.type==="number")&&!c.canRelation){let u=/^[a-z0-9-]+$/,h=$=>u.test(`${$}`),m=c.key.toLowerCase()!=="id";return!(!c.accessor||typeof c.accessor=="string")||!m?c:(c.multiple?o[c.key]:[o[c.key]]).every($=>h(c.accessor?$[c.accessor]:$))?{...c,canRelation:!0}:c}return c});var Ur=/(image|photo|avatar|cover|thumbnail|icon)/i,fr=e=>e.map(s=>s.type==="url"&&Ur.test(s.key)?{...s,type:"image"}:s);var gr=(e,r)=>e.map(o=>{let p=Re(o,r);return p?{...o,relation:!0,type:"relation",resource:p}:o});var hr=(e,r,s,o,p)=>e.map(u=>{if(u.relation&&u.type==="relation"&&!u.resource){let h=u.accessor?o[u.key][u.accessor]:o[u.key],m=p(u.key,h,o,p);if(m&&m.type!=="relation")return{...u,fieldable:!0,relation:!1,type:m.type,canRelation:!0}}return u});var xr=[fr,gr,hr,yr];var W=R(require("react")),le=R(require("prism-react-renderer")),$r=R(require("prism-react-renderer/themes/vsDark")),Pe=({code:e,wrapperProps:r})=>((0,W.useEffect)(()=>{var o;let s=document.createElement("style");s.appendChild(document.createTextNode(`
                #code-highlight::-webkit-scrollbar {-webkit-appearance: none; height: 7px; width: 7px;}
                #code-highlight::-webkit-scrollbar-thumb {background-color: #666b7a;}
                #code-highlight::-webkit-scrollbar-corner {background-color: rgb(30, 30, 30);}
                `)),(o=document.getElementById("code-highlight"))==null||o.appendChild(s)},[]),W.default.createElement("div",{id:"code-highlight",style:{maxHeight:"75vh",height:"100%",overflow:"auto",backgroundColor:"rgb(30, 30, 30)"},...r!=null?r:{}},W.default.createElement(le.default,{...le.defaultProps,theme:$r.default,code:e,language:"tsx"},({className:s,style:o,tokens:p,getLineProps:c,getTokenProps:u})=>W.default.createElement("pre",{className:s,style:{...o,padding:"14px 14px 14px 14px",margin:"0",width:"100%",boxSizing:"border-box"}},p.map((h,m)=>W.default.createElement("div",{key:m,...c({line:h,key:m})},h.map((x,F)=>W.default.createElement("span",{key:F,...u({token:x,key:F})}))))))));var K=R(require("react")),Wr=R(require("@refinedev/core")),me=require("@aliemir/react-live");var Ir=[["react","React",K.default],["@refinedev/core","RefineCore",Wr]],De=K.default.createContext({}),Oe=({fetchError:e,code:r,additionalScope:s,errorComponent:o})=>{var h;let p=(h=r&&Fe(Ae(r,[...Ir,...s!=null?s:[]].reduce((m,[x,F,k,$])=>$?m:{...m,[x]:F},{}))))!=null?h:"",c=K.default.useMemo(()=>({React:K.default,...K.default,...[...Ir,...s!=null?s:[]].reduce((m,[x,F,k])=>({...m,[F]:k}),{})}),[s]),u=K.default.useMemo(()=>()=>{let{error:x}=K.default.useContext(De);return o?K.default.createElement(o,{error:x?`<p>An error occured while rendering the generated component.You can check the generated code from the below "Show Code" button and fix the error manually.</p>
                                    <p>If you think this is a bug, please report the issue at <a target="_blank" rel="noopener noreferrer" href="https://github.com/refinedev/refine/issues">https://github.com/refinedev/refine/issues</a></p>
                                    <p>Exception:</p>
                                    <code>${x}</code>`:typeof e=="string"?e:e?`<p>Something went wrong while fetching the resource data.</p>
                                    <p>To learn more about the Inferencer, please check the <a href="https://refine.dev/docs/packages/documentation/inferencer/" target="_blank">documentation</a>.</p>`:void 0}):null},[o,e]);return K.default.createElement(me.LiveProvider,{Context:De,code:p,scope:c,noInline:!0},!e&&K.default.createElement(me.LivePreview,{Context:De}),K.default.createElement(u,null))};var J=R(require("react")),de=require("@refinedev/core");var br=(e,r,s)=>{let{resource:o,id:p,resources:c}=(0,de.useResource)(r),u=s!=null?s:p,h=(0,de.useDataProvider)(),[m,x]=J.default.useState(void 0),[F,k]=J.default.useState(void 0),[$,I]=J.default.useState(!0),[S,a]=J.default.useState(!1),w=J.default.useCallback(async T=>{var N,V;let i=(N=ae(o))!=null?N:Ie(o==null?void 0:o.name,void 0,c),b=h(i);a(!0),x(void 0);try{if((e==="list"||e==="create")&&o){let t=(V=(await b.getList({resource:o==null?void 0:o.name})).data)==null?void 0:V[0];t||x(`<p>No records/data found for resource "${o==null?void 0:o.name}".</p>
                            <p>Please check your data provider and resource.</p>
                            <p>For more info, please check the <a href="https://refine.dev/docs/packages/documentation/inferencer/" target="_blank">documentation</a>.</p>`),k(t),setTimeout(()=>{a(!1)},500)}if((e==="edit"||e==="show")&&T&&o){let t=(await b.getOne({resource:o==null?void 0:o.name,id:T})).data;t||x(`<p>No records/data found for resource "${o==null?void 0:o.name}".</p>
                            <p>Please check your data provider and resource.</p>
                            <p>For more info, please check the <a href="https://refine.dev/docs/packages/documentation/inferencer/" target="_blank">documentation</a>.</p>`),k(t),setTimeout(()=>{a(!1)},500)}}catch(O){console.warn("An error occured while fetching the resource data. Please check the error message below:",O),x(`<p>Something went wrong while fetching the resource data.</p>
                    <p>Please check your data provider and API for resource "${o==null?void 0:o.name}".</p>
                    <p>For more info, please check the <a href="https://refine.dev/docs/packages/documentation/inferencer/" target="_blank">documentation</a>.</p>`),setTimeout(()=>{a(!1)},500)}},[e,h,o,c]);return J.default.useEffect(()=>{I(!1),!S&&!F&&w(u)},[w,u]),{data:F,loading:S,initial:$,error:m}};var re=R(require("react")),kr=require("@refinedev/core");var je=require("lodash"),vr=({record:e,fields:r,infer:s})=>{let o=(0,kr.useDataProvider)(),[p,c]=re.default.useState([]),[u,h]=re.default.useState(!0),[m,x]=re.default.useState(!1),F=re.default.useCallback(async k=>{console.groupCollapsed("@refinedev/inferencer is trying to detect relations");let $=[];x(!0);try{let I=k.map(async a=>{if(a&&(a.relation||a.canRelation)&&e){let w=ae(a.resource),T=o(w),i=Array.isArray(a.accessor)?void 0:a.multiple?e[a.key].map(b=>a.accessor?(0,je.get)(b,a.accessor):b)[0]:a.accessor?(0,je.get)(e[a.key],a.accessor):e[a.key];if(i&&a.resource)try{let{data:b}=await T.getOne({resource:a.resource.name,id:i});$.push({status:"success",resource:a.resource.name,field:a.key});let N=s("__",b,{},s);return{...a,relationInfer:N}}catch{return $.push({status:"error",resource:a.resource.name,field:a.key}),{...a,relationInfer:null}}if(i){let b,N;try{let{data:t}=await T.getOne({resource:z(H(a.key)),id:i});$.push({status:"success",resource:z(H(a.key)),field:a.key}),b=t,N=!0}catch{$.push({status:"error",resource:z(H(a.key)),field:a.key});try{let{data:C}=await T.getOne({resource:D(H(a.key)),id:i});$.push({status:"success",resource:D(H(a.key)),field:a.key}),b=C,N=!1}catch{return $.push({status:"error",resource:D(H(a.key)),field:a.key}),{...a,relationInfer:null}}}let V=s("__",b,{},s),O=H(a.key);return{...a,relation:!0,type:"relation",resource:{name:N?z(O):D(O)},fieldable:!1,canRelation:void 0,relationInfer:V}}return{...a,relationInfer:null}}return a}),S=await Promise.all(I);c(S.filter(a=>a)),setTimeout(()=>{x(!1)},500)}catch{setTimeout(()=>{x(!1)},500)}setTimeout(()=>{console.log(`Tried to detect relations with ${$.length} attempts and succeeded with ${$.filter(I=>I.status==="success").length} attempts.`),console.groupEnd()},500)},[o,e]);return re.default.useEffect(()=>{h(!1),!m&&r&&r.length>0&&p.length===0&&F(r)},[F,m,r,F]),{fields:p,loading:m,initial:u}};var Z=({type:e,additionalScope:r=[],customElements:s=[],fieldTransformers:o=[],renderer:p,loadingComponent:c,errorComponent:u,codeViewerComponent:h})=>{let m=er([...dr,...s]),x=rr([...xr,...o]),F=({resourceName:$,fieldTransformer:I,id:S})=>{var n,y,E,j,U,ne;let{resource:a,resources:w}=(0,He.useResource)($),{resource:T}=(0,He.useResource)(),{data:i,loading:b,initial:N,error:V}=br(e,$!=null?$:a==null?void 0:a.name,S),O=_.default.useMemo(()=>{if(i){let Fr=Object.keys(i).map(oe=>{let Be=i[oe];return m(oe,Be,i,m)}).filter(Boolean);if(a){let oe=x(Fr,w,a,i,m);return I?oe.flatMap(qe=>{let Ue=I(qe);return Ue?[Ue]:[]}):oe}return[]}return[]},[i,w,a,I]),{fields:t,loading:C}=vr({record:i,fields:O,infer:m}),L=_.default.useMemo(()=>!b&&!C&&!N&&a?p({resource:a,resources:w,fields:t,infer:m,isCustomPage:a.name!==(T==null?void 0:T.name),id:S}):"",[a,w,t,b,C]);return _.default.createElement(_.default.Fragment,null,c&&(b||C)&&_.default.createElement(c,null),!b&&!C&&_.default.createElement(_.default.Fragment,null,_.default.createElement(Oe,{fetchError:!b&&V?V:!b&&!N&&!i,code:we(L,M((ne=(U=(j=(E=(n=a==null?void 0:a.meta)==null?void 0:n.label)!=null?E:(y=a==null?void 0:a.options)==null?void 0:y.label)!=null?j:a==null?void 0:a.label)!=null?U:a==null?void 0:a.name)!=null?ne:"Resource",e)),errorComponent:u,additionalScope:r}),h&&_.default.createElement(h,{code:Se(L),loading:b||C})))};return({name:$,resource:I,fieldTransformer:S,id:a})=>_.default.createElement(F,{fieldTransformer:S,resourceName:I!=null?I:$,key:I!=null?I:$,id:a})};var ye=R(require("react")),G=({error:e})=>e?ye.default.createElement("div",{style:{display:"flex",justifyContent:"center",alignItems:"center",minHeight:"300px",position:"relative"}},ye.default.createElement("div",{style:{position:"absolute",top:"50%",left:"50%",transform:"translate(-50%, -50%)"}},ye.default.createElement("div",{style:{fontSize:"1.5rem",fontWeight:"bold"},dangerouslySetInnerHTML:{__html:e!=null?e:""}}))):null;var te=R(require("react")),Y=()=>{let[e,r]=te.default.useState(0);return te.default.useEffect(()=>{if(typeof window<"u"){let s=setInterval(()=>{r(o=>o===3?0:o+1)},300);return()=>{clearInterval(s)}}return()=>{}},[]),te.default.createElement("div",{style:{display:"flex",justifyContent:"center",alignItems:"center",minHeight:"120px",position:"relative"}},te.default.createElement("div",{style:{position:"absolute",top:"50%",left:"50%",transform:"translate(-50%, -50%)"}},te.default.createElement("span",{style:{fontSize:"1rem",fontWeight:600}},"loading",".".repeat(e))))};var v=R(require("react")),Q=require("@tabler/icons");var Zr=({visible:e,onClose:r,children:s})=>e?v.default.createElement("div",{style:{position:"fixed",top:0,left:0,right:0,bottom:0,zIndex:9999,backgroundColor:"rgba(0, 0, 0, 0.5)"}},v.default.createElement("div",{style:{position:"absolute",top:"50%",left:"50%",transform:"translate(-50%, -50%)",width:"80%",height:"80%",backgroundColor:"#fff",maxWidth:"700px",borderRadius:"4px",padding:"16px",paddingTop:"32px"}},v.default.createElement("div",{style:{position:"absolute",top:4,right:6,width:"32px",height:"32px",fontSize:"24px",lineHeight:"32px",cursor:"pointer",textAlign:"center"},onClick:r},"\xD7"),s)):null,X=({code:e,loading:r})=>{let s=v.default.useMemo(()=>ke(e!=null?e:""),[e]),o=v.default.useRef(null),[p,c]=v.default.useState(!1),[u,h]=v.default.useState(!1),m=v.default.useCallback(()=>{h(!0)},[]),x=v.default.useCallback(()=>{h(!1)},[]);if(r)return null;let F=()=>{var k,$,I,S;(k=o==null?void 0:o.current)==null||k.select(),($=o==null?void 0:o.current)==null||$.setSelectionRange(0,Number.MAX_SAFE_INTEGER),typeof navigator<"u"&&(navigator.clipboard.writeText((S=(I=o==null?void 0:o.current)==null?void 0:I.value)!=null?S:""),c(!0),setTimeout(()=>{c(!1)},1e3))};return s?v.default.createElement(v.default.Fragment,null,v.default.createElement("div",{style:{position:"fixed",bottom:"12px",right:"64px"}},v.default.createElement("div",{style:{display:"flex",flexDirection:"row",alignItems:"center",justifyContent:"center",width:"100%",height:"100%",gap:"12px"}},v.default.createElement("button",{style:{backgroundColor:"#fff",border:"1px solid #e2e8f0",borderRadius:"4px",padding:"8px 16px",cursor:"pointer",display:"flex",fontFamily:"sans-serif",alignItems:"center"},onClick:m},v.default.createElement(Q.IconCode,{size:18}),v.default.createElement("span",{style:{marginLeft:"8px",fontSize:"14px",color:"#666B7A"}},"Show Code")),v.default.createElement("a",{href:"https://github.com/refinedev/refine/discussions/3046",target:"_blank",rel:"noreferrer",style:{backgroundColor:"#fff",border:"1px solid #e2e8f0",borderRadius:"4px",padding:"8px 16px",cursor:"pointer",display:"flex",alignItems:"center",textDecoration:"none",color:"#666B7A",fontFamily:"sans-serif"}},v.default.createElement(Q.IconMessageCircle,{size:18}),v.default.createElement("span",{style:{marginLeft:"8px",fontSize:"14px",color:"#666B7A"}},"Give Feedback")))),v.default.createElement("div",{style:{height:"50px"}}),v.default.createElement(Zr,{visible:u,onClose:x},v.default.createElement("div",{style:{display:"flex",flexDirection:"column",alignItems:"center",justifyContent:"space-between",width:"100%",height:"100%"}},v.default.createElement("div",{style:{display:"flex",flexDirection:"row",alignItems:"center",justifyContent:"center",width:"100%",height:"100%"}},v.default.createElement(Pe,{code:s,wrapperProps:{style:{width:"100%",maxHeight:"75vh",height:"100%",overflow:"auto",backgroundColor:"rgb(30, 30, 30)"}}}),v.default.createElement("textarea",{ref:o,defaultValue:(s!=null?s:"").replace(/\\n/g,`\r
`),id:"code-input",style:{padding:0,margin:0,width:0,height:0,opacity:0,border:"none",display:"block"}})),v.default.createElement("div",{style:{display:"flex",flexDirection:"column",alignItems:"flex-end",justifyContent:"center",paddingTop:"8px",width:"100%",height:"100%"}},v.default.createElement("button",{style:{backgroundColor:"#fff",border:"1px solid #e2e8f0",borderRadius:"4px",padding:"8px 16px",cursor:"pointer",display:"flex",alignItems:"center"},onClick:F},p?v.default.createElement(Q.IconCheck,{size:18}):v.default.createElement(Q.IconCopy,{size:18}),v.default.createElement("span",{style:{marginLeft:"8px",fontSize:"14px",color:"#666B7A"}},p?"Copied":"Copy")))))):null};var ze=({resource:e,fields:r,isCustomPage:s,id:o})=>{var t,C,L;let p=M((t=e.label)!=null?t:e.name,"show"),c="record",u=[["React","react",!0],["useShow","@refinedev/core"],["useResource","@refinedev/core"],["useNavigation","@refinedev/core"]],m=r.filter(n=>(n==null?void 0:n.relation)&&!(n!=null&&n.fieldable)&&(n==null?void 0:n.resource)).filter(Boolean).map(n=>{if((n==null?void 0:n.relation)&&!n.fieldable&&n.resource){if(n.multiple){u.push(["useMany","@refinedev/core"]);let y=l(c,n.key);return n.accessor&&(y=`${l(c,n.key)}?.map((item: any) => ${l("item",void 0,n.accessor)})`),`
                const { data: ${A(n.key,"Data")}, isLoading: ${A(n.key,"IsLoading")} } =
                useMany({
                    resource: "${n.resource.name}",
                    ids: ${y} || [],
                    queryOptions: {
                        enabled: !!${c},
                    },
                });
                `}return u.push(["useOne","@refinedev/core"]),`
                const { data: ${A(n.key,"Data")}, isLoading: ${A(n.key,"IsLoading")} } =
                useOne({
                    resource: "${n.resource.name}",
                    id: ${l(c,n.key,n.accessor,!1)} || "",
                    queryOptions: {
                        enabled: !!${c},
                    },
                });
            `}}).filter(Boolean),x=n=>{if(n.relation&&n.resource){let y=A(n.key,"Data"),E=A(n.key,"IsLoading");return n.multiple?g`
                <div style={{ marginTop: "6px" }}>
                    <h5>${f(n.key)}</h5>
                    <ul>
                    {${E} ? <>Loading...</> : (
                        <>
                        ${(()=>{var j;if(n.relationInfer)if((j=n.relationInfer)!=null&&j.accessor){if(Array.isArray(n.relationInfer.accessor))return"";{let U=A(n.key),ne=l(U,void 0,n.relationInfer.accessor);return g`
                                            {${y}?.data?.map((${U}: any) => <li key={${ne}}>{${ne}}</li>)}
                                        `}}else return"";else return""})()}
                        </>
                    )}
                    </ul>
                </div>
                `:g`
                <div style={{ marginTop: "6px" }}>
                    <h5>${f(n.key)}</h5>
                    <div>
                    {${E} ? <>Loading...</> : (
                        <>
                        ${(()=>{var j;return n.relationInfer?(j=n.relationInfer)!=null&&j.accessor?Array.isArray(n.relationInfer.accessor)?`{${l(`${y}?.data`,void 0,n.relationInfer.accessor,' + " " + ')}}`:`{${y}?.data?.${n.relationInfer.accessor}}`:`{${y}?.data}`:`{${y}?.data?.id}`})()}
                        </>
                    )}
                    </div>
                </div>
                `}},F=n=>{if(n.type==="text"){if(n.multiple){let y=l("item",void 0,n.accessor);return g`
                <div style={{ marginTop: "6px" }}>
                    <h5>${f(n.key)}</h5>
                    <ul>
                        {${l(c,n.key)}?.map((item: any) => (
                            <li key={${y}}>
                                {${y}}
                            </li>
                        ))}
                    </ul>
                </div>
            `}return g`
                <div style={{ marginTop: "6px" }}>
                    <h5>${f(n.key)}</h5>
                    <div>
                    {${l(c,n.key,n.accessor)}}
                    </div>
                </div>
            `}},k=n=>{if(n.type==="image"){if(n.multiple){let y=l("item",void 0,n.accessor);return g`
                <div style={{ marginTop: "6px" }}>
                    <h5>${f(n.key)}</h5>
                    <ul>
                    {${l(c,n.key)}?.map((item: any) => (
                        <li key={${y}}>
                            <img style={{ maxWidth: 200 }} src={${y}}/>
                        </li>
                    ))}
                    </ul>
                </div>
            `}return g`
                <div style={{ marginTop: "6px" }}>
                    <h5>${f(n.key)}</h5>
                    <img src={{ maxWidth: 200 }} src={${l(c,n.key,n.accessor," + ")}} />
                </div>
                `}},$=n=>{if(n.type==="email"){if(n.multiple){let y=l("item",void 0,n.accessor);return g`
                <div style={{ marginTop: "6px" }}>
                    <h5>${f(n.key)}</h5>
                    <ul>
                        {${l(c,n.key)}?.map((item: any) => (
                            <li key={${y}}>
                                <a href={"mailto:" + ${y}}>{${y}}</a>
                            </li>
                        ))}
                    </ul>
                </div>
            `}return g`
                <div style={{ marginTop: "6px" }}>
                    <h5>${f(n.key)}</h5>
                    <div>
                        <a href={"mailto:" + ${l(c,n.key,n.accessor," + ")}}>
                            {${l(c,n.key,n.accessor," + ")}}
                        </a>
                    </div>
                </div>
            `}},I=n=>{if(n.type==="url"){if(n.multiple){let y=l("item",void 0,n.accessor);return g`
                <div style={{ marginTop: "6px" }}>
                    <h5>${f(n.key)}</h5>
                    <ul>
                        {${l(c,n.key)}?.map((item: any) => (
                            <li key={${y}}>
                                <a href={${y}}>{${y}}</a>
                            </li>
                        ))}
                    </ul>
                </div>
            `}return g`
            <div style={{ marginTop: "6px" }}>
                <h5>${f(n.key)}</h5>
                <div>
                    <a href={${l(c,n.key,n.accessor," + ")}} target="_blank" rel="noreferrer">
                        {${l(c,n.key,n.accessor," + ")}}
                    </a>
                </div>
            </div>
            `}},S=n=>{if(n.type==="boolean"){if(n.multiple){let y=l("item",void 0,n.accessor);return g`
                <div style={{ marginTop: "6px" }}>
                    <h5>${f(n.key)}</h5>
                    <ul>
                        {${l(c,n.key)}?.map((item: any) => (
                            <li key={${y}}>
                                {${y} ? "Yes" : "No"}
                            </li>
                        ))}
                    </ul>
                </div>
            `}return g`
            <div style={{ marginTop: "6px" }}>
                <h5>${f(n.key)}</h5>
                <div>
                {${l(c,n.key,n.accessor," && ")} ? "Yes" : "No"}
                </div>
            </div>
            `}},a=n=>{if(n.type==="date"){if(n.multiple){let y=l("item",void 0,n.accessor);return g`
                <div style={{ marginTop: "6px" }}>
                    <h5>${f(n.key)}</h5>
                    <ul>
                    {${l(c,n.key)}?.map((item: any) => (
                        <li key={${y}}>
                            {(new Date(${y})).toLocaleString(undefined, { timeZone: "UTC" })}
                        </li>
                    ))}
                    </ul>
                </div>
            `}return g`
            <div style={{ marginTop: "6px" }}>
                <h5>${f(n.key)}</h5>
                <div>
                    {(new Date(${l(c,n.key,n.accessor," + ' ' + ")})).toLocaleString(undefined, { timeZone: "UTC" })}
                </div>
            </div>
            `}},w=n=>{if(n.type==="richtext")return g`
            <div style={{ marginTop: "6px" }}>
                <h5>${f(n.key)}</h5>
                <p>
                {${l(c,n.key,n.accessor,' + " " + ')}}
                </p>
            </div>
            `},T=n=>{if(n.type==="number"){if(n.multiple){let y=l("item",void 0,n.accessor);return g`
                <div style={{ marginTop: "6px" }}>
                    <h5>${f(n.key)}</h5>
                    <ul>
                        {${l(c,n.key)}?.map((item: any) => (
                        <li key={${y}}>
                            {${y}}
                        </li>
                        ))}
                    </ul>
                </div>
                `}return g`
                <div style={{ marginTop: "6px" }}>
                    <h5>${f(n.key)}</h5>
                    <div>
                    {${l(c,n.key,n.accessor,' + " " + ')} ?? ""}
                    </div>
                </div>
            `}},i=n=>{if(n)return g`
                ${n}
            `},b=r.map(n=>{switch(n==null?void 0:n.type){case"text":return i(F(n));case"number":return i(T(n));case"richtext":return i(w(n));case"email":return i($(n));case"image":return i(k(n));case"date":return i(a(n));case"boolean":return i(S(n));case"url":return i(I(n));case"relation":return i(x(n));default:return}}),{canEdit:N,list:V}=e!=null?e:{},O=!!V;return q(u),g`
    ${B(u)}
    
    export const ${p} = () => {
        const { edit, list } = useNavigation();
        ${s?"":"const { id } = useResource();"}
        const { queryResult } = useShow(${s?`{ 
                    resource: "${e.name}", 
                    id: ${o}
                }`:""});
        const { data, isLoading } = queryResult;

        const ${c} = data?.data;
    
        ${m}

        return (
            <div style={{ padding: "16px" }}>
            <div style={{ display: "flex", alignItems: "center", justifyContent: "space-between" }}>
                <h1>
                ${f(D((C=e.label)!=null?C:e.name))}
                </h1>
                <div style={{ display: "flex", gap: "8px" }}>
                ${O?g`<button onClick={() => list("${e.name}")}>${f(z((L=e.label)!=null?L:e.name)+" list")}</button>`:""}
                ${N?g`<button onClick={() => edit("${e.name}", ${s?`"${o}"`:"id ?? ''"})}>Edit</button>`:""}
                </div>
            </div>
                <div>
                    ${b.join("")}
                </div>
            </div>
        );
    };
    `},fe=Z({type:"show",additionalScope:[],codeViewerComponent:X,loadingComponent:Y,errorComponent:G,renderer:ze});var Gr=R(require("@refinedev/react-table")),Yr=R(require("@tanstack/react-table"));var ee=e=>Array.isArray(e.accessor)||e.multiple?`accessorKey: "${e.key}"`:e.accessor?`accessorKey: "${P(e.key,void 0,e.accessor)}"`:`accessorKey: "${e.key}"`,Ke=({resource:e,fields:r,isCustomPage:s})=>{var V,O;let o=M((V=e.label)!=null?V:e.name,"list"),p="tableData?.data",c=[["React","react",!0],["IResourceComponentsProps","@refinedev/core"],["useNavigation","@refinedev/core"],["useTable","@refinedev/react-table"],["ColumnDef","@tanstack/react-table"],["flexRender","@tanstack/react-table"]],u=r.filter(t=>(t==null?void 0:t.relation)&&!(t!=null&&t.fieldable)&&(t==null?void 0:t.resource)),h=u.filter(Boolean).map(t=>{if((t==null?void 0:t.relation)&&!t.fieldable&&t.resource){c.push(["GetManyResponse","@refinedev/core"]),c.push(["useMany","@refinedev/core"]);let C="";return t.multiple?C=`[].concat(...(${p}?.map((item) => ${l("item",t.key,t.accessor,!1)}) ?? []))`:C=`${p}?.map((item) => ${l("item",t.key,t.accessor,!1)}) ?? []`,`
                const { data: ${A(t.key,"Data")} } =
                useMany({
                    resource: "${t.resource.name}",
                    ids: ${C},
                    queryOptions: {
                        enabled: !!${p},
                    },
                });
                `}}).filter(Boolean),m=u==null?void 0:u.map(t=>{if(t&&t.resource)return A(t.key,"Data")}).filter(Boolean),x=t=>{var C;if(t.relation&&t.resource){let L=`${A(t.key,"Data")}?.data`;if(Array.isArray(t.accessor))return;let n=`id: "${t.key}"`,y=`header: "${f(t.key)}"`,E=ee(t),j="";if(t.multiple){let U="item";t!=null&&t.relationInfer&&(U=l("item",void 0,t.relationInfer.accessor)),j=`cell: function render({ getValue, table }) {
                    const meta = table.options.meta as {
                        ${A(t.key,"Data")}: GetManyResponse;
                    };

                    ${t!=null&&t.accessor?"try {":""}

                    const ${A(t.key,"")} = getValue<any[]>()?.map((item) => {
                        return meta.${A(t.key,"Data")}?.data?.find(
                            (resourceItems) => resourceItems.id === ${l("item",void 0,t.accessor)}
                        );
                    })


                    return (
                        <ul>
                            {${A(t.key,"")}?.map((item, index) => (
                                <li key={index}>
                                    {${U}}
                                </li>
                            ))}
                        </ul>
                    )
                    ${t!=null&&t.accessor?" } catch (error) { return null; }":""}
                }
            `}else t!=null&&t.relationInfer?j=`cell: function render({ getValue, table }) {
                        const meta = table.options.meta as {
                            ${A(t.key,"Data")}: GetManyResponse;
                        };

                        ${t!=null&&t.accessor?"try {":""}

                        const ${A(t.key,"")} = meta.${L}?.find(
                            (item) => item.id === getValue<any>(),
                        );

                        return ${l(A(t.key),void 0,(C=t==null?void 0:t.relationInfer)==null?void 0:C.accessor)} ?? "Loading...";

                        ${t!=null&&t.accessor?" } catch (error) { return null; }":""}
                    },`:j="";return`
                {
                    ${n},
                    ${y},
                    ${E},
                    ${j}
                }
            `}},F=t=>{if(t.type==="image"){let C=`id: "${t.key}"`,L=ee(t),n=`header: "${f(t.key)}"`,y=g`
                cell: function render({ getValue }) {
                    ${t!=null&&t.accessor?"try {":""}
                        return <img style={{ maxWidth: "100px" }} src={${l("getValue<any>()",void 0,Array.isArray(t.accessor)?t.accessor:void 0," + ")}} />
                    ${t!=null&&t.accessor?" } catch (error) { return null; }":""}
                }
            `;if(t.multiple){let E=l("item",void 0,t.accessor," + ");y=`
                    cell: function render({ getValue }) {
                        ${t!=null&&t.accessor?"try {":""}
                            return (
                                <ul>
                                    {getValue<any[]>()?.map((item, index) => (
                                        <li key={index}><img src={${E}} style={{ height: "50px", maxWidth: "100px" }} /></li>
                                    ))}
                                </ul>
                            )
                        ${t!=null&&t.accessor?" } catch (error) { return null; }":""}
                    }
                `}return`
                {
                    ${C},
                    ${L},
                    ${n},
                    ${y}
                }
            `}},k=t=>{if(t.type==="email"){let C=`id: "${t.key}"`,L=ee(t),n=`header: "${f(t.key)}"`,y=g`
                cell: function render({ getValue }) {
                    return <a href={"mailto:" + ${l("getValue<any>()",void 0,Array.isArray(t.accessor)?t.accessor:void 0,' + " " + ')}}>{${l("getValue<any>()",void 0,Array.isArray(t.accessor)?t.accessor:void 0,' + " " + ')}}</a>
                }
            `;return t.multiple&&(y=`
                    cell: function render({ getValue }) {
                        return (
                            <ul>
                                {getValue<any[]>()?.map((item, index) => (
                                    <li key={index}>
                                        {${l("item",void 0,t.accessor," + ")}}
                                    </li>
                                ))}
                            </ul>
                        )
                    }
                `),`
                {
                    ${C},
                    ${L},
                    ${n},
                    ${y}
                }
            `}},$=t=>{if(t.type==="url"){let C=`id: "${t.key}"`,L=ee(t),n=`header: "${f(t.key)}"`,y=g`
                cell: function render({ getValue }) {
                    return <a href={${l("getValue<any>()",void 0,Array.isArray(t.accessor)?t.accessor:void 0," + ")}}>{${l("getValue<any>()",void 0,Array.isArray(t.accessor)?t.accessor:void 0," + ")}}</a>
                }
            `;return t.multiple&&(y=`
                    cell: function render({ getValue }) {
                        return (
                            <ul>
                                {getValue<any[]>()?.map((item, index) => (
                                    <li key={index}>
                                        {${l("item",void 0,t.accessor," + ")}}
                                    </li>
                                ))}
                            </ul>
                        )
                    }
                `),`
                {
                    ${C},
                    ${L},
                    ${n},
                    ${y}
                }
            `}},I=t=>{if((t==null?void 0:t.type)==="boolean"){let C=`id: "${t.key}"`,L=ee(t),n=`header: "${f(t.key)}"`,y=g`
                cell: function render({ getValue }) {
                    return ${l("getValue<any>()",void 0,Array.isArray(t.accessor)?t.accessor:void 0," + ")} ? "yes" : "no"
                }
            `;return t.multiple&&(y=`
                    cell: function render({ getValue }) {
                        return (
                            <ul>
                                {getValue<any[]>()?.map((item, index) => (
                                    <li key={index}>
                                        {${l("item",void 0,t.accessor," + ")} ? "yes" : "no"}
                                    </li>
                                ))}
                            </ul>
                        );
                    }
                `),`
                {
                    ${C},
                    ${L},
                    ${n},
                    ${y}
                }
            `}},S=t=>{if(t.type==="date"){let C=`id: "${t.key}"`,L=ee(t),n=`header: "${f(t.key)}"`,y=g`
                cell: function render({ getValue }) {
                    return (new Date(${l("getValue<any>()",void 0,Array.isArray(t.accessor)?t.accessor:void 0,' + " " + ')})).toLocaleString(undefined, { timeZone: "UTC" })
                }
            `;return t.multiple&&(y=`
                    cell: function render({ getValue }) {
                        return (
                            <ul>
                                {getValue<any[]>()?.map((item, index) => (
                                    <li key={index}>
                                    {(new Date(${l("item",void 0,t.accessor," + ")})).toLocaleString(undefined, { timeZone: "UTC" })}
                                    </li>
                                ))}
                            </ul>
                        )
                    }
                `),`
                {
                    ${C},
                    ${L},
                    ${n},
                    ${y}
                }
            `}},a=t=>{if(t&&(t.type==="text"||t.type==="number"||t.type==="richtext")){let C=`id: "${t.key}"`,L=ee(t),n=`header: "${f(t.key)}"`,y="";return t.multiple&&(y=`
                    cell: function render({ getValue }) {
                        return (
                            <ul>
                                {getValue<any[]>()?.map((item, index) => (
                                    <li key={index}>
                                        {${l("item",void 0,t.accessor,' + " " + ')}}
                                    </li>
                                ))}
                            </ul>
                        )
                    }
                `),!t.multiple&&Array.isArray(t.accessor)&&(y=`
                    cell: function render({ getValue }) {
                        return (
                            <>{${l("getValue<any>()",t.key,t.accessor)}}</>
                        );
                    }
                `),`
                {
                    ${C},
                    ${L},
                    ${n},
                    ${y}
                }
            `}},{canEdit:w,canShow:T,canCreate:i}=e!=null?e:{},b=w||T?g`
    {
        id: "actions",
        accessorKey: "id",
        header: "Actions",
        cell: function render({ getValue }) {
            return (
                <div
                    style={{
                        display: "flex",
                        flexDirection: "row",
                        flexWrap: "wrap",
                        gap: "4px",
                    }}
                >
                ${T?g`
                    <button
                        onClick={() => {
                            show("${e.name}", getValue() as string);
                        }}
                    >
                        Show
                    </button>
                    `:""}
                    ${w?g`
                            <button
                            onClick={() => {
                                edit("${e.name}", getValue() as string);
                            }}
                        >
                            Edit
                        </button>
                    `:""}
                </div>
            );
        },
    },
        `:"",N=r.map(t=>{switch(t==null?void 0:t.type){case"text":case"number":case"richtext":return a(t);case"email":return k(t);case"image":return F(t);case"date":return S(t);case"boolean":return I(t);case"url":return $(t);case"relation":return x(t);default:return}});return q(c),g`
    ${B(c)}
    
    export const ${o}: React.FC<IResourceComponentsProps> = () => {
        const columns = React.useMemo<ColumnDef<any>[]>(() => [
            ${[...N,b].filter(Boolean).join(",")}
        ], []);

        ${w||T?g`
        const { ${w?"edit,":""} ${T?"show,":""} ${i?"create,":""} } = useNavigation();
        `:""}

        const {
            getHeaderGroups,
            getRowModel,
            setOptions,
            refineCore: {
                tableQueryResult: { data: tableData },
            },
            getState,
            setPageIndex,
            getCanPreviousPage,
            getPageCount,
            getCanNextPage,
            nextPage,
            previousPage,
            setPageSize,
            getColumn,
        } = useTable({
            columns,
            ${s?`
            refineCoreProps: {
                resource: "${e.name}",
            }
            `:""}
            
        });

        ${h}

        setOptions((prev) => ({
            ...prev,
            meta: {
                ...prev.meta,
                ${m.join(", ")}
            },
        }));

        return (
            <div style={{ padding: "16px" }}>
            <div style={{ display: "flex", alignItems: "center", justifyContent: "space-between" }}>
                <h1>${f((O=e.label)!=null?O:e.name)}</h1>
                ${i?g`<button onClick={() => create("${e.name}")}>Create</button>`:""}
            </div>
                <div style={{ maxWidth: "100%", overflowY: "scroll" }}>
                    <table>
                        <thead>
                            {getHeaderGroups().map((headerGroup) => (
                                <tr key={headerGroup.id}>
                                    {headerGroup.headers.map((header) => (
                                        <th key={header.id}>
                                            {!header.isPlaceholder && (
                                                flexRender(
                                                    header.column.columnDef
                                                        .header,
                                                    header.getContext(),
                                                )
                                            )}
                                        </th>
                                    ))}
                                </tr>
                            ))}
                        </thead>
                        <tbody>
                            {getRowModel().rows.map((row) => (
                                <tr key={row.id}>
                                    {row.getVisibleCells().map((cell) => (
                                        <td key={cell.id}>
                                            {flexRender(
                                                cell.column.columnDef.cell,
                                                cell.getContext(),
                                            )}
                                        </td>
                                    ))}
                                </tr>
                            ))}
                        </tbody>
                    </table>
                </div>
                <div style={{ marginTop: "12px" }}>
                    <button
                        onClick={() => setPageIndex(0)}
                        disabled={!getCanPreviousPage()}
                    >
                        {"<<"}
                    </button>
                    <button
                        onClick={() => previousPage()}
                        disabled={!getCanPreviousPage()}
                    >
                        {"<"}
                    </button>
                    <button onClick={() => nextPage()} disabled={!getCanNextPage()}>
                        {">"}
                    </button>
                    <button
                        onClick={() => setPageIndex(getPageCount() - 1)}
                        disabled={!getCanNextPage()}
                    >
                        {">>"}
                    </button>
                    <span>
                        Page
                        <strong>
                            {getState().pagination.pageIndex + 1} of{" "}
                            {getPageCount()}
                        </strong>
                    </span>
                    <span>
                        | Go to page:
                        <input
                            type="number"
                            defaultValue={getState().pagination.pageIndex + 1}
                            onChange={(e) => {
                                const page = e.target.value
                                    ? Number(e.target.value) - 1
                                    : 0;
                                setPageIndex(page);
                            }}
                        />
                    </span>{" "}
                    <select
                        value={getState().pagination.pageSize}
                        onChange={(e) => {
                            setPageSize(Number(e.target.value));
                        }}
                    >
                        {[10, 20, 30, 40, 50].map((pageSize) => (
                            <option key={pageSize} value={pageSize}>
                                Show {pageSize}
                            </option>
                        ))}
                    </select>
                </div>
            </div>   
        );
    };
    `},ge=Z({type:"list",additionalScope:[["@refinedev/react-table","RefineReactTable",Gr],["@tanstack/react-table","TanstackReactTable",Yr]],codeViewerComponent:X,loadingComponent:Y,errorComponent:G,renderer:Ke});var Qr=R(require("@refinedev/core")),Xr=R(require("@refinedev/react-hook-form"));var Me=({resource:e,fields:r,isCustomPage:s})=>{var $,I,S;let o=M(($=e.label)!=null?$:e.name,"create"),p=[["React","react",!0],["useNavigation","@refinedev/core"],["useForm","@refinedev/react-hook-form"]],u=r.filter(a=>(a==null?void 0:a.relation)&&!(a!=null&&a.fieldable)&&(a==null?void 0:a.resource)).filter(Boolean).map(a=>{if((a==null?void 0:a.relation)&&!a.fieldable&&a.resource)return p.push(["useSelect","@refinedev/core"]),`
                const { options: ${A(a.key,"Options")} } =
                useSelect({
                    resource: "${a.resource.name}",
                    ${ce(a)}
                });
            `}).filter(Boolean),h=a=>{if(a.relation&&a.resource){p.push(["useSelect","@refinedev/core"]);let w=A(a.key,"Options");return g`
            <label>
                <span style={{ marginRight: "8px" }}>
                    ${f(a.key)}
                </span>
                <select
                    placeholder="Select ${D(a.resource.name)}"
                    {...register("${P(a.key,void 0,a.accessor)}", {
                        required: ${a.multiple?"false":'"This field is required"'},
                    })}
                >
                    {${w}?.map((option) => (
                        <option value={option.value} key={option.value}>
                            {option.label}
                        </option>
                    ))}
                </select>
                <span style={{ color: "red" }}>
                    {${l("(errors as any)",a.key,a.accessor,!1)}?.message as string}
                </span>
            </label>
            `}},m=a=>{if(a.type==="text"||a.type==="url"||a.type==="email"||a.type==="number"||a.type==="date"||a.type==="richtext"){if(ie(a.key)||a.multiple)return;let w=a.type==="richtext"?"textarea":"input";return g`
                <label>
                    <span style={{ marginRight: "8px" }}>
                        ${f(a.key)}
                    </span>
                    <${w}
                        ${a.type!=="date"&&a.type!=="richtext"?`type="${a.type}"`:""}
                        ${w==="textarea"?`
                            rows={5}
                            cols={33}
                            style={{ verticalAlign: "top" }}
                        `:""}
                        {...register("${P(a.key,void 0,a.accessor)}", {
                            required: "This field is required",
                            ${a.type==="number"?"valueAsNumber: true,":""}
                        })}

                    />
                    <span style={{ color: "red" }}>
                        {${l("(errors as any)",a.key,a.accessor,!1)}?.message as string}
                    </span>
                </label>
            `}},x=a=>{if(a.type==="boolean")return a.multiple?void 0:g`
                <label>
                    <span style={{ marginRight: "8px" }}>
                        ${f(a.key)}
                    </span>
                    <input
                        type="checkbox"
                        {...register("${P(a.key,void 0,a.accessor)}", {
                            required: "This field is required",
                        })}
                    />
                    <span style={{ color: "red" }}>
                        {${l("errors",a.key,a.accessor,!1)}?.message as string}
                    </span>
                </label>
            `},F=r.map(a=>{switch(a==null?void 0:a.type){case"text":case"number":case"email":case"url":case"richtext":case"date":return m(a);case"boolean":return x(a);case"relation":return h(a);default:return}}),k=!!e.list;return q(p),g`
    ${B(p)}
    
    export const ${o} = () => {
        ${k?`
        const { list } = useNavigation();
        `:""}
        const {
            refineCore: { onFinish, formLoading },
            register,
            handleSubmit,
            resetField,
            formState: { errors },
        } = useForm(
            ${s?`
            { 
                refineCoreProps: {
                    resource: "${e.name}",
                    action: "create",
                }
            }`:""}
        );
    
        ${u}

        return (
            <div style={{ padding: "16px" }}>
                <div style={{ display: "flex", justifyContent: ${k?'"space-between"':'"flex-start"'} }}>
                    <h1>
                        ${f(D((I=e.label)!=null?I:e.name)+" Create")}
                    </h1>
                    ${k?g`
                            <div>
                        <button
                                onClick={() => {
                                    list("${e.name}");
                                }}
                        >
                            ${f(z((S=e.label)!=null?S:e.name)+" List")}
                        </button>
                        </div>
                    `:""}
                </div>
                <form onSubmit={handleSubmit(onFinish)}>
                    <div style={{ display: "flex", flexDirection: "column", gap: "8px" }}>
                        ${F.join("")}
                        <div>
                            <input type="submit" value="Save" />
                        </div>
                    </div>
                </form>
            </div>
        );
    };
    `},he=Z({type:"create",additionalScope:[["@refinedev/core","RefineCore",Qr],["@refinedev/react-hook-form","RefineReactHookForm",Xr]],codeViewerComponent:X,loadingComponent:Y,errorComponent:G,renderer:Me});var Jr=R(require("@refinedev/core")),et=R(require("@refinedev/react-hook-form")),rt=R(require("react-hook-form"));var _e=({resource:e,fields:r,isCustomPage:s,id:o})=>{var S,a,w,T;let p=M((S=e.label)!=null?S:e.name,"edit"),c=A((a=e.label)!=null?a:e.name,"Data"),u=[["React","react",!0],["useNavigation","@refinedev/core"],["useForm","@refinedev/react-hook-form"]],m=r.filter(i=>(i==null?void 0:i.relation)&&!(i!=null&&i.fieldable)&&(i==null?void 0:i.resource)).filter(Boolean).map(i=>{if((i==null?void 0:i.relation)&&!i.fieldable&&i.resource){u.push(["useSelect","@refinedev/core"]);let b=l(c,i.key,i.accessor,!1);return i.multiple&&i.accessor&&(b=`${l(c,i.key)}?.map((item: any) => ${l("item",void 0,i.accessor)})`),`
                const { options: ${A(i.key,"Options")} } =
                useSelect({
                    resource: "${i.resource.name}",
                    defaultValue: ${b},
                    ${ce(i)}
                });

                React.useEffect(() => {
                    resetField("${P(i.key,void 0,i.accessor)}");
                }, [${A(i.key,"Options")}]);
            `}}).filter(Boolean),x=i=>{if(i.relation&&i.resource){u.push(["useSelect","@refinedev/core"]);let b=A(i.key,"Options");return g`
            <label>
                <span style={{ marginRight: "8px" }}>${f(i.key)}</span>
                <select
                    placeholder="Select ${D(i.resource.name)}"
                    {...register("${P(i.key,void 0,i.accessor)}", {
                        required: ${i.multiple?"false":'"This field is required"'},
                    })}
                >
                    {${b}?.map((option) => (
                        <option value={option.value} key={option.value}>
                            {option.label}
                        </option>
                    ))}
                </select>
                <span style={{ color: "red" }}>
                    {${l("(errors as any)",i.key,i.accessor,!1)}?.message as string}
                </span>
            </label>
            `}},F=i=>{if(i.type==="text"||i.type==="url"||i.type==="email"||i.type==="number"||i.type==="date"||i.type==="richtext"){if(i.multiple){let N=P(i.key,"${index}",i.accessor),V=l(l("(errors as any)",i.key,void 0,!1)+"?.[index]",void 0,i.accessor);return`
                    <>
                        {${l(c,i.key)}?.map((item: any, index: number) => (
                            <label key={index}>
                                <span style={{ marginRight: "8px" }}>
                                    ${f(i.key)} #{index + 1}
                                </span>
                                <input ${i.type!=="richtext"?`
                                type="${i.type!=="date"?i.type:"text"}"
                                `:""} {...register(\`${N}\`, { required: "This field is required", ${i.type==="number"?"valueAsNumber: true,":""} })} />
                                <span style={{color: "red"}}>
                                {${l(V,"message")} as string}
                                </span>
                            </label>
                        ))}
                    </>
                `}let b=i.type==="richtext"?"textarea":"input";return g`
            <label>
                <span style={{ marginRight: "8px" }}>${f(i.key)}</span>
                <${b}
                ${ie(i.key)?"disabled":""}
                ${i.type!=="date"&&i.type!=="richtext"?`type="${i.type}"`:""}
                ${b==="textarea"?`
                    rows={5}
                    cols={33}
                    style={{ verticalAlign: "top" }}
                `:""}
                {...register("${P(i.key,void 0,i.accessor)}", {
                    required: "This field is required",
                    ${i.type==="number"?"valueAsNumber: true,":""}
                })}
                />
                <span style={{ color: "red" }}>
                {${l("(errors as any)",i.key,i.accessor,!1)}?.message as string}
                </span>
            </label>
            `}},k=i=>{if(i.type==="boolean"){if(i.multiple){let b=P(i.key,void 0,i.accessor),N=l("(errors as any)",i.key,void 0,!1)+"?.[index]";return`
                    <>
                        {${l(c,i.key)}?.map((item: any, index: number) => (
                            <label key={index}>
                            <span style={{ marginRight: "8px" }}>
                            ${f(i.key)} #{index + 1}
                            </span>
                            <input
                                type="checkbox"
                                {...register(\`${b}.\${index}\`, {
                                    required: "This field is required",
                                })}
                            />
                            <span style={{ color: "red" }}>
                                {${N}?.message as string}
                            </span>
                        ))}
                    </>
                `}return g`
            <label>
                <span style={{ marginRight: "8px" }}>${f(i.key)}</span>
                <input type="checkbox" {...register("${P(i.key,void 0,i.accessor)}", {
                    required: "This field is required",
                })} />
                <span style={{ color: "red" }}>
                    {${l("errors",i.key,i.accessor,!1)}?.message as string}
                </span>
            </label>
            `}},$=r.map(i=>{switch(i==null?void 0:i.type){case"text":case"number":case"email":case"url":case"richtext":case"date":return F(i);case"boolean":return k(i);case"relation":return x(i);default:return}}),I=!!e.list;return q(u),g`
    ${B(u)}
    
    export const ${p} = () => {
        ${I?`
        const { list } = useNavigation();
        `:""}
        const {
            refineCore: { onFinish, formLoading, queryResult },
            register,
            handleSubmit,
            resetField,
            formState: { errors },
        } = useForm(
            ${s?`
            { 
                refineCoreProps: {
                    resource: "${e.name}",
                    id: ${o},
                    action: "edit",
                }
            }`:""}
        );
    
        const ${c} = queryResult?.data?.data;
    
        ${m}

        return (
            <div style={{ padding: "16px" }}>
                <div style={{ display: "flex", justifyContent: ${I?'"space-between"':'"flex-start"'} }}>
                    <h1>
                        ${f(D((w=e.label)!=null?w:e.name)+" Edit")}
                    </h1>
                    ${I?g`
                    <div>
                        <button
                                onClick={() => {
                                    list("${e.name}");
                                }}
                        >
                            ${f(z((T=e.label)!=null?T:e.name)+" List")}
                        </button>
                    </div>
                    `:""}
                </div>
                <form onSubmit={handleSubmit(onFinish)}>
                    <div style={{ display: "flex", flexDirection: "column", gap: "8px" }}>
                        ${$.join("")}
                        <div>
                            <input type="submit" value="Save" />
                        </div>
                    </div>
                </form>
            </div>
        );
    };
    `},xe=Z({type:"edit",additionalScope:[["@refinedev/core","RefineCore",Jr],["@refinedev/react-hook-form","RefineReactHookForm",et],["react-hook-form","ReactHookForm",rt]],codeViewerComponent:X,loadingComponent:Y,errorComponent:G,renderer:_e});var tt=({action:e,id:r,...s})=>{let{action:o,id:p}=(0,Cr.useResource)();switch(e!=null?e:o){case"show":return pe.default.createElement(fe,{...s,id:r!=null?r:p});case"create":return pe.default.createElement(he,{...s,id:r!=null?r:p});case"edit":return pe.default.createElement(xe,{...s,id:r!=null?r:p});default:return pe.default.createElement(ge,{...s,id:r!=null?r:p})}};
